File: field_force/field_force/page/app_user_attendance_report/app_user_attendance_report.py
import frappe
import json

from field_force.field_force.page.utils import generate_excel_and_download, get_time_in_12_hour_format
from field_force.field_force.report.utils import set_image_url, set_link_to_doc, get_site_directory_path, \
    set_google_map_location_button

@frappe.whitelist()
def get_user_attendance_data(filters=None):
    columns = get_columns()
    query_result = get_query_data(filters)
    site_directory = get_site_directory_path()

    for index, app_user_attendance in enumerate(query_result):
        set_link_to_doc(app_user_attendance, 'sales_person', 'sales-person')
        app_user_attendance.server_time = get_time_in_12_hour_format(app_user_attendance.server_time)
        app_user_attendance.device_time = get_time_in_12_hour_format(app_user_attendance.device_time)

        app_user_attendance['name'] = f'<a href="/app/app-user-attendance/{app_user_attendance.name}" ' \
                                      f'target="_blank">{app_user_attendance.name}</a>'
        # app_user_attendance.server_date = f"{app_user_attendance.server_date}<br>{app_user_attendance.server_time}"
        # app_user_attendance.device_date = f"{app_user_attendance.device_date}<br>{app_user_attendance.device_time}"

        app_user_attendance.cheated = 'Yes' if app_user_attendance.cheated else 'No'
        app_user_attendance.location = set_google_map_location_button(app_user_attendance)
        set_image_url(app_user_attendance, site_directory)
        app_user_attendance['sl'] = index + 1

    # data = query_result
    return query_result, columns

def get_query_data(filters):
    try:
        filters = json.loads(filters)
    except:
        pass
    conditions = get_conditions(filters)

    query_string = '''select app_user_attendance.name, app_user_attendance.server_date, app_user_attendance.server_time,
                    app_user_attendance.type, app_user_attendance.device_date, app_user_attendance.device_time, 
                    app_user_attendance.latitude, app_user_attendance.longitude, app_user_attendance.device_model, 
                    app_user_attendance.image, app_user_attendance.sales_person, app_user_attendance.cheated
                    from `tabApp User Attendance` app_user_attendance where %s 
                    order by app_user_attendance.server_date desc, app_user_attendance.server_time desc''' % (conditions)

    query_result = frappe.db.sql(query_string, as_dict=1, debug=0)
    return query_result


def get_conditions(filters):
    from_date = filters.get('from_date')
    to_date = filters.get('to_date')
    sales_person = filters.get('sales_person')
    type = filters.get('type')

    conditions = []

    if type:
        conditions.append('app_user_attendance.type = "%s"' % type)
    if from_date:
        conditions.append('app_user_attendance.server_date >= "%s"' % from_date)
    if to_date:
        conditions.append('app_user_attendance.server_date <= "%s"' % to_date)
    if sales_person:
        conditions.append('app_user_attendance.sales_person = "%s"' % sales_person)

    return " and ".join(conditions)

def get_columns():
    columns =  [
        {'fieldname': 'sl', 'label': 'SL', 'expwidth': 5, 'export': False},
        {'fieldname': 'server_date', 'label': 'Date', 'expwidth': 12},
        {'fieldname': 'server_time', 'label': 'Time', 'expwidth': 12},
        {'fieldname': 'name', 'label': 'ID', 'expwidth': 15},
        {'fieldname': 'sales_person', 'label': 'Sales Person', 'expwidth': 15},
        {'fieldname': 'type', 'label': 'Type', 'expwidth': 10},
        {'fieldname': 'device_date', 'label': 'Device Date', 'expwidth': 15},
        {'fieldname': 'device_time', 'label': 'Device Time', 'expwidth': 15},
        {'fieldname': 'cheated', 'label': 'Cheated', 'expwidth': 10},
        {'fieldname': 'location', 'label': 'Location', 'expwidth': 15},
        {'fieldname': 'image', 'label': 'Image', 'fieldtype': 'Image', 'expwidth': 15, 'export': False}
    ]
    return columns

@frappe.whitelist()
def export_file(**filters):
    columns = get_columns()
    data = get_export_data(filters)
    file_name = 'App_User_Attendance_Report.xlsx'
    generate_excel_and_download(columns, data, file_name)

def get_export_data(filters):
    query_result = get_query_data(filters)

    for index, app_user_attendance in enumerate(query_result):
        # app_user_attendance.user = app_user_attendance.user_fullname or app_user_attendance.user
        # app_user_attendance.server_date = f"{app_user_attendance.server_date}\n{app_user_attendance.server_time}"
        # app_user_attendance.device_date = f"{app_user_attendance.device_date}\n{app_user_attendance.device_time}"
        app_user_attendance.cheated = 'Yes' if app_user_attendance.cheated else 'No'
        app_user_attendance['sl'] = index + 1

    return query_result

File: field_force/field_force/page/app_user_attendance_report/app_user_attendance_report.js
frappe.pages['app-user-attendance-report'].on_page_load = (wrapper) => {
    frappe.require("assets/field_force/css/page_modal.css");
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'App User Attendance Report',
        single_column: true
    });

    window.addEventListener("click", function(event) {
      if (event.target.id !== "image") {
        document.getElementById('modal_section').style.display = 'none';
      }
    });

    $('.page-body').css('background', '#FFFFFF');
    new AppUserAttendanceReport(page);
};


class AppUserAttendanceReport {
    constructor(page) {
        this.page = page;
        this.make_form();
        this.make_menu();
        this.initialize_modal();
    }

    make_menu = () => {
        this.page.add_action_icon("refresh", () => {
            this.fetch_and_render();
        });
        this.page.add_menu_item("Export", ()=>{
            this.export_excel()
        })
    }
    make_form = () => {
        this.form = new frappe.ui.FieldGroup({
            fields: [
                {
                    fieldname: 'from_date',
                    label: __('From Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                    reqd: 1,
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'to_date',
                    label: __('To Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'sales_person',
                    label: __('Sales Person'),
                    fieldtype: 'Link',
                    options: 'Sales Person',
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'type',
                    label: __('Type'),
                    fieldtype: 'Select',
                    options: [
                        "",
                        {"value": "Checkin", "label": __("Check In")},
                        {"value": "Checkout", "label": __("Check Out")},
                    ],
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Section Break'
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'preview'
                }
            ],
            body: this.page.body
        });
        this.form.make();
    }
    initialize_modal = () => {
      let html = `<div id="modal_section" class="modal-2">
          <div class="modal-image">
          <img style="transform: rotate(-90deg) " class="modal-content-2" id="img" alt="img">
          </div>
        </div>`
        // this.form.body.append(html);
        $('.main-section').append(html)
    }
    fetch_and_render = () => {
        let {from_date, to_date, sales_person, type} = this.form.get_values();
        if (!from_date) {
            this.form.get_field('preview').html('');
            return;
        }
        this.form.get_field('preview').html(`
        	<div class="text-muted margin-top">
        		${__("Fetching...")}
        	</div>
        `);
        frappe.call('field_force.field_force.page.app_user_attendance_report.app_user_attendance_report.get_user_attendance_data', {
            filters: {
                from_date: from_date,
                to_date: to_date,
                sales_person: sales_person,
                type: type
            },
            freeze: true

        }).then(r => {
            let diff = r.message[0];
            let fields = r.message[1];
            this.render(diff, fields);
        });
    }

    render = (diff, fields) => {
        let table_header = this.table_header(fields);
        let table_body = this.table_body(diff, fields);
        this.form.get_field('preview').html(`<table class="table table-bordered" id="export_excel">${table_header}${table_body}</table>`);
    }

    // table_header = () => {
    //     let table_header = '<thead>\n' +
    //         '    <tr>\n' +
    //         '      <th scope="col" >SL</th>\n' +
    //         '      <th scope="col">DateTime</th>\n' +
    //         // '      <th scope="col">Time</th>\n' +
    //         '      <th scope="col">ID</th>\n' +
    //         '      <th scope="col">User</th>\n' +
    //         '      <th scope="col">Type</th>\n' +
    //         '      <th scope="col">Device DateTime</th>\n' +
    //         '      <th scope="col">Cheated</th>\n' +
    //         // '      <th scope="col">Device Time</th>\n' +
    //         '      <th scope="col">Latitude</th>\n' +
    //         '      <th scope="col">Longitude</th>\n' +
    //         '      <th scope="col">Model</th>\n' +
    //         '      <th scope="col">Image</th>\n' +
    //         '    </tr>\n' +
    //         '  </thead>\n';
    //     return table_header;

    table_header = (headers) => {
        let table_header = `<thead><tr>`;

        headers.forEach(function (data, index){
            table_header += `<th scope="col">${data.label || ''}</th>`;
        })

        table_header += `</tr>`;
        return table_header;
    }

    table_body = (diff, fields) => {
        var html = `<tbody>`;

        diff.forEach(function (data, index) {
            html += `<tr>`;

            fields.forEach(function (field, index){
               html += get_absolute_format_and_html(field, data[field.fieldname]);
            })

            html += `</tr>`;
        })

        html += `</tbody>`;
        return html
    }

    // table_body = (diff) => {
    //     var html = "<tbody>";
    //     diff.forEach(function (data, index) {
    //         html += `<tr>`;
    //         html += '<td>' + data.sl + '</td>';
    //         html += '<td>' + `${data.server_date || ''}` + '</td>';
    //         // html += '<td>' + `${data.server_time || ''}` + '</td>';
    //         html += '<td>' + `${data.name || ''}` + '</td>';
    //         html += '<td>' + `${data.user || ''}` + '</td>';
    //         html += '<td>' + `${data.type || ''}` + '</td>';
    //         html += '<td>' + `${data.device_date || ''}` + '</td>';
    //         // html += '<td>' + `${data.device_time || ''}` + '</td>';
    //         html += '<td>' + `${data.cheated || ''}` + '</td>';
    //         html += '<td>' + `${data.latitude || ''}` + '</td>';
    //         html += '<td>' + `${data.longitude || ''}` + '</td>';
    //         html += '<td>' + `${data.device_model || ''}` + '</td>';
    //         html += '<td style="height:100px; width:120px;"><a href="#"><img style="height:100%; width:100%" src="' + data.image + '" onclick="(function(e){document.getElementById(\'modal_section\').style.display=\'block\';document.getElementById(\'img01\').src=e.path[0].currentSrc;return false;})(arguments[0]);return false;"></a></td>';
    //         html += `</tr>`;
    //     })
    //     html += "</tbody>";
    //     return html
    // }

    export_excel = () => {
        let {from_date, to_date, sales_person, type} = this.form.get_values();
        let url = `/api/method/field_force.field_force.page.app_user_attendance_report.app_user_attendance_report.export_file`;
        url += `?from_date=${from_date||''}&to_date=${to_date||''}&sales_person=${sales_person||''}&type=${type||''}`;
        window.open(url, '_blank');
    }
}

function get_absolute_format_and_html(field, value){
    if (field.fieldtype === "Image"){
        return get_image_html(value);
    }
    else if (field.fieldtype === "Currency"){
        return get_currency_format(value);
    }
    else if (field.fieldtype ==="Data") {
        return `<td>${value || ''}</td>`;
    }
    else {
        return `<td>${value || ''}</td>`;
    }
}
function get_image_html(image_url) {
    let rotate = "transform: rotate(-90deg)"

    if (image_url === "/files/default-image.png"){
        rotate = '';
    }

    return `
        <td style="height:100px; width:120px;">
            <a href="#">
                <img style="height:100%; width:100%; ${rotate}" src="${image_url}" id="image" img-path="${image_url}" onclick="(
                    function(e){
                        document.getElementById(\'modal_section\').style.display=\'block\';
                        var nAgt = navigator.userAgent;
                        if (nAgt.indexOf('Safari') !== -1) {
                            document.getElementById(\'img\').src=e.target.currentSrc;
                        }else{
                            document.getElementById(\'img\').src=e.path[0].currentSrc;
                        }
                        return false;
                    }
                )(arguments[0]);
                return false;">
            </a>
        </td>`;
}

function get_currency_format(value){
    const company = frappe.defaults.get_user_default("company");
    const currency = frappe.get_doc(":Company", company).default_currency;
    let value_in_currency = format_currency(value, currency);
    return `<td>${value_in_currency || ''}</td>`
}
